Add support for DESCRIBE INPUT

rschlussel opened the pull request:
Syntax:
DESCRIBE INPUT <statement_name>

Describes the requirements for the input parameters to a prepared
statement.  For each expected parameter, a row will be returned
containing the position of the parameter and the expected type if
it can be determined, or unknown if not.

Set operations and insert queries don't resolve types until planning, so we won't be able to get proper types for those queries.  If this is the way to go for describe input, I can file a ticket to address that after everything else goes in (it's rather complicated to change the relation coercions to happen at analysis time).

electrum commented:
This makes `INPUT` a reserved word. Should it be non-reserved?

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:
```diff
@@ -96,12 +96,16 @@
 
-    public Analysis(Statement root, List<Expression> parameters)
+    // for describe input and describe output
+    private final boolean isDescribe;
```
This should probably be `isDescribeInput` so it's not confused with `DESCRIBE` / `SHOW COLUMNS`

rschlussel replied:
it also gets used for describe output in the next PR, but `isDescribeInputOrOutput` seems too long.  Any ideas for a name that encompasses both?

rschlussel commented:
You're right.  I'll add `INPUT` to the non-reserved keywords

rschlussel commented:
ping.  I made 'input' non-reserved.

dain commented:
This needs docs.  Also do we want this to be `DESCRIBE INPUT` or do we want this to be a `SHOW` command to match the others?

dain commented:
@martint and @electrum can you comment on the syntax.  I can handle the rest of this PR.

dain commented:
Also is there an `information_schema` table for prepared statement inputs?

rschlussel commented:
Describe input comes from the SQL spec  (was reviewed at https://github.com/prestodb/presto/issues/4338, though the implementation is different). 

I'm not sure what you mean by information_schema table.  Prepared statements get passed along like session variables.  The purpose of describe input is to analyze the prepared statement and tell you the number and types of parameter inputs required to execute.

rschlussel commented:
@martint can you confirm the syntax please so that this can be reviewed?

martint commented:
Yes, DESCRIBE INPUT is good.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeInputRewrite.java:
```diff
@@ -0,0 +1,145 @@
+            // get all parameters in query
+            List<Parameter> parameters = getParameters(statement);
+
+            // return the positions and types of all parameters.  If there are no parameters, return a single null row.
```
Is `return a single null row` a spec defined behavior

rschlussel replied:
No. I initially did it that way because there was no way to put 0 rows into a values statement because then you don't know what your fields are, and a null row seemed to sufficiently convey the same information.  I'm realizing now that I can just add a limit 0 for a case with no parameters.

electrum replied:
What does the specification say for this?

rschlussel replied:
The spec doesn't have very much detail, but generally for SQL when there are no results you return no rows.
This is what the specification says:
>To provide the necessary information to applications, SQL permits an application to request the SQL-server to describe a prepared statement. The description of a statement identifies the number of input dynamic parameters (describe input) and their data type information or it identifies the number of output dynamic parameters or values to be returned (describe output) and their data type information.

dain commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -7062,4 +7062,34 @@ public void testParametersNonPreparedStatement()
+        MaterializedResult expected = resultBuilder(session, BIGINT, VARCHAR)
+                .row(0, "unknown")
+                .row(1, "bigint")
+                .row(2, "varchar(0)")
```
is `varchar(0)` really correct here?

rschlussel replied:
Probably not, since that then gets cast to varchar(20).  Maybe for parametrized types we should just get the base name.

dain commented on presto-docs/src/main/sphinx/sql/describe-input.rst:
```diff
@@ -0,0 +1,50 @@
+       -----------------
+       0        | unknown
+       1        | bigint
+       2        | varchar
```
this doesn't match up with the test

dain commented:
@martint other than my comments, the execution code and docs look good.  Can you review the analyzer changes?  Specifically, the use of the `UNKNOWN` type.

martint commented:
Looks good. I'll merge it after the current release goes out.

electrum commented on presto-docs/src/main/sphinx/sql/describe-input.rst:
```diff
@@ -0,0 +1,49 @@
+Examples
+--------
+
+1. Describe query with 3 parameters:
```
Using a numbered list here is a bit strange. Let's write this section like this:
```
Examples
--------

Prepare and describe a query with three parameters:

.. code-block:: sql

    PREPARE my_select FROM
    SELECT ? FROM nation WHERE regionkey = ? AND name < ?

.. code-block:: sql
    
    DESCRIBE my_select

.. code-block:: none

     Position | Type
    --------------------
            0 | unknown
            1 | bigint
            2 | varchar
    (3 rows)

Describe a query with no parameters:

electrum commented on presto-docs/src/main/sphinx/sql/execute.rst:
```diff
@@ -0,0 +1,48 @@
+.. _execute:
```
Remove

electrum commented on presto-docs/src/main/sphinx/sql/execute.rst:
```diff
@@ -0,0 +1,48 @@
+
+.. code-block:: none
+
+    EXECUTE statement_name [USING parameter1[, parameter2, ...]]
```
Brackets should have surrounding spaces in the grammar (see SELECT for an example). Write it list this:
```
EXECUTE statement_name [ USING value [, ...] ]

electrum commented on presto-docs/src/main/sphinx/sql/execute.rst:
```diff
@@ -0,0 +1,48 @@
+-----------
+
+Executes a prepared statement with the name ``statement_name``. Parameter values
+are defined in the USING clause. For more information on preparing a statement
```
Same "See Also"

electrum replied:
Use formatting for `USING`

electrum commented on presto-docs/src/main/sphinx/sql/execute.rst:
```diff
@@ -0,0 +1,48 @@
+Examples
+--------
+
+* Execute a query with no parameters:
```
Follow the example I gave for `DESCRIBE INPUT`

electrum commented on presto-docs/src/main/sphinx/sql/prepare.rst:
```diff
@@ -0,0 +1,46 @@
+.. _prepare:
```
Same comments

electrum commented on presto-docs/src/main/sphinx/sql/prepare.rst:
```diff
@@ -0,0 +1,46 @@
+
+Prepare a select query::
+
+    PREPARE my_select1
```
Keep the `FROM` on the previous line since it's part of the `PREPARE` and shouldn't be mixed with the query

electrum replied:
```
PREPARE my_select1 FROM
SELECT * FROM nation

electrum commented on presto-docs/src/main/sphinx/sql/prepare.rst:
```diff
@@ -0,0 +1,46 @@
+Prepare a select query::
+
+    PREPARE my_select1
+    FROM SELECT * from NATION;
```
Capitalize keywords like `FROM` and lowercase table names and other identifiers

electrum commented on presto-docs/src/main/sphinx/sql/prepare.rst:
```diff
@@ -0,0 +1,46 @@
+    FROM SELECT * from NATION;
+
+Prepare a select query that includes parameters. The values to compare with
+regionkey and nationkey will be filled in with the :ref:`execute` statement::
```
Use code formatting for table names

electrum replied:
Use `:doc:` instead of `:ref:`

electrum commented:
The documentation needs a lot of changes.

electrum commented:
It might be helpful if you can split out the documentation to a separate pull request.

rschlussel commented:
Addressed all comments on the code, squashed them in and rebased on latest master.
@electrum I pulled out the documentation from this pr.  I'll address your comments and open a new pr for the docs.

electrum commented:
Thanks! Assigning back to @martint

martint commented:
Merged, thanks!
